Add Developer guide for further development. (#147)
sunpy/sunpy-soar: docs/dev_guide/tables.rst

.. _sunpy-soar-dev-guide-tables:

**********************************
Tables supported in ``sunpy-soar``
**********************************

The ``sunpy-soar`` library currently supports data retrieval from both science and low-latency data tables, such as ``v_sc_data_item`` and ``v_ll_data_item``.

> Additionally, it provides support for join tables associated with remote instruments, such as ``v_eui_sc_fits``.

Reviewer: Done on next line.

.. _sunpy-soar-dev-guide-tables:

**********************************
Tables supported in ``sunpy-soar``
**********************************

> The ``sunpy-soar`` library currently supports data retrieval from both science and low-latency data tables, such as
> ``v_sc_data_item`` and ``v_ll_data_item``. Additionally, it provides support for join tables associated with remote
> instruments, such as ``v_eui_sc_fits``.

Reviewer 2: ```suggestion
The ``sunpy-soar`` library currently supports data retrieval from both science and low-latency data tables, such as ``v_sc_data_item`` and ``v_ll_data_item``.
Additionally, it provides support for join tables associated with remote instruments, such as ``v_eui_sc_fits``.
```
Should we not explain what the difference between these three style of tables are?

> In the context of ``sunpy-soar``, data tables contain columns related to scientific measurements, while instrument
> tables contain metadata specific to particular instruments. The ``sunpy-soar`` library specifically supports the
> wavelength and detector columns within these tables. These columns are linked to the data columns in the instrument
> tables through a join operation, using the ``data_item_oid`` as the key.

Reviewer 2: ```suggestion
In the context of ``sunpy-soar``, data tables contain columns related to scientific measurements, while instrument tables contain metadata specific to particular instruments.
The ``sunpy-soar`` library specifically supports the wavelength and detector columns within these tables.
These columns are linked to the data columns in the instrument tables through a join operation, using the ``data_item_oid`` as the key.
```

How can a new column be added?
==============================

> If the new column you wish to add is part of an existing data or instrument FITS table, you can extend the
> corresponding ADQL query to include this column. For example, if the new column is in the instrument table, it should
> be included in the SELECT clause of the query for that table.

Reviewer 2: ```suggestion
If the new column you wish to add is part of an existing data or instrument FITS table, you can extend the corresponding ADQL query to include this column.
For example, if the new column is in the instrument table, it should be included in the "SELECT" clause of the query for that table.
```

> Moreover, if filtering by this new column is required, you must consider adding it as an attribute in the ``attrs.py``
> file. If the column already exists within `sunpy.net.attrs`, you should add a corresponding walker in the
> ``attrs.py`` file. The ``_can_handle_query`` method in the ``client.py`` file should also be updated to ensure proper
> support for this new column in queries.

Reviewer 2: ```suggestion
Moreover, if one needs to enable filtering by this new column, you must consider adding it as an attribute in the ``attrs.py`` file.
If the column already exists within `sunpy.net.attrs`, you should add a corresponding walker in the ``attrs.py`` file.
The ``_can_handle_query`` method in the ``client.py`` file should also be updated to ensure proper support for this new column in queries.
```

As above, some linking to files and an example adding to the ``_can_handle_query`` would be helpful.

How can a new table be added?
=============================

> When adding support for a new table in ``sunpy-soar``, it is essential to fully understand the context and requirements.
> This includes identifying the key that will be used for join operations and determining what data should be
> returned to the user.

Reviewer 2: ```suggestion
When adding support for a new table in ``sunpy-soar``, it is essential to fully understand the context and requirements.
This includes identifying the key that will be used for join operations and determining what data should be returned to the user.
```
Does data here mean columns?

Author: yes, I'll use column.

> For direct joins, the key relationship between the tables should be carefully identified. Once the joins are
> established, the necessary columns can be included in the query, following the process outlined for adding columns.

Reviewer 2: ```suggestion
For direct joins, the key relationship between the tables should be carefully identified.
Once the joins are established, the necessary columns can be included in the query, following the process outlined for adding columns.
```

> Depending on the use case, you might need to determine the type of join (e.g., inner join, outer join, left join) and what
> specific tables or columns should be displayed to the user. In some cases, conditional logic may be required, where
> certain attributes in the search query trigger specific join operations or column selections.

Reviewer 2: ```suggestion
Depending on the use case, you might need to determine the type of join (e.g., inner join, outer join, left join) and what specific tables or columns should be displayed to the user.
In some cases, conditional logic may be required, where certain attributes in the search query trigger specific join operations or column selections.
```

Any examples of the conditional logic?

Author: okay, I'll add.

Author: The explanation I provided was specifically in the context of handling joins with the FoV (Field of View) table, where certain columns are returned when the user includes FoV in their search query. However, since the implementation details of this feature are still under discussion, it might be best to omit this section for now.

Once the FoV pull request (PR) is finalized and the implementation becomes clear, I am eager to update the guide with the new capabilities and features that will be introduced.

> Finally, ensure that the ``_do_search`` method is updated to reflect these changes. This method should handle any
> additional columns or conditional logic for attribute-specific queries, ensuring that the appropriate data is returned
> to the user.

Reviewer 2: ```suggestion
Finally, ensure that the ``_do_search`` method is updated to reflect these changes. 
This method should handle any additional columns or conditional logic for attribute-specific queries, ensuring that the appropriate data is returned to the user.
```
Can we not link to the do_search method?

> Note
> ====
> 
> Accordingly, test cases for every extreme and usual case must be added, to show full functionality of the added table or column.

Reviewer 2: ```suggestion
```

.. _sunpy-soar-dev-guide-tables:

**********************************
Tables supported in ``sunpy-soar``
**********************************

The ``sunpy-soar`` library currently supports data retrieval from both science and low-latency data tables, such as ``v_sc_data_item`` and ``v_ll_data_item``.
Additionally, it provides support for join tables associated with remote instruments, such as ``v_eui_sc_fits``.

> These tables and their columns are described in the `Tables, Views, and Columns <https://www.cosmos.esa.int/web/soar/tables-views-and-columns/>`_.

Reviewer 2: ```suggestion
These tables and their columns are described in the `Tables, Views, and Columns <https://www.cosmos.esa.int/web/soar/tables-views-and-columns/>`__.
```

How can a new table be added?
=============================

When adding support for a new table in ``sunpy-soar``, it is essential to fully understand the context and requirements.
This includes identifying the key that will be used for join operations and determining what columns should be returned to the user.

For direct joins, the key relationship between the tables should be carefully identified.
Once the joins are established, the necessary columns can be included in the query, following the process outlined for adding columns.

Depending on the use case, you might need to determine the type of join (e.g., inner join, outer join, left join) and what specific tables or columns should be displayed to the user.

Finally, ensure that the ``_do_search`` method is updated to reflect these changes.
This method should handle any additional columns or conditional logic for attribute-specific queries, ensuring that the appropriate columns is returned to the user.

Allowing an attribute to work in the search query
=================================================

For any attribute to work in the search query, we can divide the situation in two cases:

> case 1: It is already there in `sunpy.net.attrs`, in this situation just a walker needs to be created.

Reviewer 2: ```suggestion
Case 1 - It is already there in `sunpy.net.attrs`, in this situation just a walker needs to be created.
```

Example on its implementation:

.. code-block:: python

    # Adding a walker applier for that attribute.
    @walker.add_applier(a.AttrName)
    def _(wlk, attr, params):
        params.append(f"Detector='{attr.value}'")

    # Depending upon the attribute the value being added could be a range of values also.
    @walker.add_applier(a.AttrName)
    def _(wlk, attr, params)
        attrmin = attr.min.value
        attrmax = attr.max.value
        # appending the attribute depending upon how it should be used in the query
        params.append(f"Attrmin='{attrmin}'+AND+Attrmax='{attrmax}'")

> case 2: It is not already there in `sunpy.net.attrs`, in this situation we have to introduce the attribute and a walker needs to be created.

Reviewer 2: ```suggestion
Case 2 - It is not already there in `sunpy.net.attrs`, in this situation we have to introduce the attribute and a walker needs to be created.
```
